mount/mount.go

```diff
@@ -71,3 +69,9 @@ type Stat struct {
 	Exists bool
 	Size   uint64
 }
+
+// MountFactory allows instantiation of a Mount from the deterministic URL representation defined by the Mount.
+type MountFactory interface {
+	// Parse initializes the mount from a URL.
+	Parse(u *url.URL) (Mount, error)
+}
```

Comment:
I'm OK going this direction for now, but I think it introduces unnecessary complexity.

What I had in mind was using reflection and the type registry to instantiate an empty struct of whichever kind, and then calling `Parse` on it to populate it. This is a bit reminiscent to JSON unmarshalling in Go.

Simplifying just for the purposes of illustration:

```
var registry map[string]reflect.Type // registry, where the key is the scheme.
var url *url.URL // url we want to parse

typ, _ := registry[url.Scheme]
mnt := reflect.New(typ).Interface().(Mount)
mnt.Parse(url) // mnt consumes the URL to populate itself
```

Reply:
The reason I suggested using the factory is so that you can inject an `api` object into the Mount.
This allows the calling code (in this case lotus) to abstract away the API call mechanism (HTTP, JSON-RPC etc) so that dagstore doesn't need to know about it.

Reply:
Gotcha. I think the underlying requirement that you're identifying is that there needs to be a way to supply some environmental context to the mount. These cannot be persisted in the URL because they may change across restarts (e.g. the miner JSON-RPC endpoint changing). The mount factory is one way to encapsulate the context but I think we'll find it too inflexible. For example, what if the sealing cluster JSON-RPC API endpoint changes at runtime? I think we'll need something more powerful going forward. But happy to keep it like this for now.

Reply:
https://github.com/filecoin-project/dagstore/issues/13